horovod/run/run_func.py

```diff
@@ -0,0 +1,199 @@
+                                   "return value: {}".format(type(e)), e)
+
+    pickled_proc_fn_str = cloudpickle.dumps(wrapped_proc_fn)
+    local_pickled_proc_fn_path = os.path.join(wdir, _LOCAL_PICKLED_PROC_FN_FILENAME)
+    with open(local_pickled_proc_fn_path, 'wb') as f:
+        f.write(pickled_proc_fn_str)
+
+    local_launcher_path = os.path.join(wdir, _LOCAL_PROC_LAUNCHER_FILENAME)
+    with open(local_launcher_path, 'w') as f:
+        f.write(textwrap.dedent("""
+        set -e
+        cd {wdir}
+        rank=${{OMPI_COMM_WORLD_RANK:-0}}
+        {exec_path} -c "import cloudpickle; cloudpickle.load(open('{fn_file}', 'rb'))(rank=$rank)"
+        """.format(wdir=wdir,
+                   exec_path=sys.executable,
+                   fn_file=_PICKLED_PROC_FN_FILENAME)))
+
+    if host is None:
+        host = "localhost:{np}".format(np=num_proc)
+
+    all_host_names = set(x.split(':')[0] for x in host.split(','))
+
+    # copy pickled function and launcher script to all hosts via scp
+    # TODO: parallelize the scp copy
+    for hostname in all_host_names:
+        scp_cmd = textwrap.dedent(
+            """
+            ssh {ssh_port_opt} -o StrictHostKeyChecking=no {hostname} mkdir -p {wdir} && \
+            scp -r -q {scp_port_opt} -o StrictHostKeyChecking=no \
+            {wdir}/{local_fn_file} {hostname}:{wdir}/{fn_file} && \
+            scp -r -q {scp_port_opt} -o StrictHostKeyChecking=no \
+            {wdir}/{local_launcher} {hostname}:{wdir}/{launcher}
+            """).format(ssh_port_opt=ssh_port_opt, scp_port_opt=scp_port_opt,
+                        wdir=wdir, hostname=hostname,
+                        local_fn_file=_LOCAL_PICKLED_PROC_FN_FILENAME,
+                        fn_file=_PICKLED_PROC_FN_FILENAME,
+                        local_launcher=_LOCAL_PROC_LAUNCHER_FILENAME,
+                        launcher=_PROC_LAUNCHER_FILENAME)
+
+        output = six.StringIO()
+        exit_code = safe_shell_exec.execute(scp_cmd, stdout=output, stderr=output)
+        if exit_code != 0:
+            output_msg = output.getvalue()
+            raise RuntimeError("Copy pickled function and launcher script to host "
+                               "{hostname} failed. Error message is {msg}."
+                               .format(hostname=hostname, msg=output_msg))
+
+    cmd = ["horovodrun", "-np", str(num_proc)]
+    if ssh_port:
+        cmd += ["-p", str(ssh_port)]
+    if host:
+        cmd += ["-H", str(host)]
+    if disable_cache:
+        cmd += ["--disable-cache"]
+    if start_timeout:
+        cmd += ["--start-timeout", str(start_timeout)]
+    if verbose:
+        cmd += ["--verbose"]
+
+    cmd += ["bash", os.path.join(wdir, _PROC_LAUNCHER_FILENAME)]
+
+    if verbose:
+        print("Run command: " + " ".join(cmd))
+
+    if env is None:
+        env = os.environ
+
+    # close stdin for sub-process, avoid sub-process to read data from stdin.
+    subprocess.check_call(cmd, stdin=subprocess.DEVNULL, env=env)
```

Comment:
Sorry, I'm wrong about check_call would capture stdout/stderr in notebook env. We still need:

~~~
p = subprocess.Popen(cmd, stdin=DEVNULL, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, env=env)
for line in p.stdout:
  sys.stdout.write(line)
p.wait()
~~~

Reply:
See my comments above.